File: tests/independent_clients/ts/call_convert.mjs
#!/usr/bin/env node
/**
 * Independent TypeScript/Node MCP client against a Control-backed Capability
 * Gateway Streamable HTTP endpoint. No shared package filesystem; only URL +
 * bearer token from the host process.
 */
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";

const endpoint = process.env.A3S_GATEWAY_ENDPOINT;
const token = process.env.A3S_GATEWAY_TOKEN;
if (!endpoint || !token) {
  console.error("A3S_GATEWAY_ENDPOINT and A3S_GATEWAY_TOKEN are required");
  process.exit(2);
}

const transport = new StreamableHTTPClientTransport(new URL(endpoint), {
  requestInit: {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  },
});
const client = new Client({ name: "independent-ts", version: "1.0.0" });
await client.connect(transport);

const listed = await client.listTools();
const names = (listed.tools ?? []).map((tool) => tool.name);
if (names.length !== 1 || names[0] !== "convert") {
  console.error(`unexpected tools: ${JSON.stringify(names)}`);
  process.exit(1);
}

const result = await client.callTool({
  name: "convert",
  arguments: { args: [] },
});
if (result.isError) {
  console.error(`call_tool failed: ${JSON.stringify(result)}`);
  process.exit(1);
}
const exitCode = result.structuredContent?.exitCode;
if (exitCode !== 0) {
  console.error(`unexpected exitCode: ${JSON.stringify(result.structuredContent)}`);
  process.exit(1);
}

await client.close();
console.log("ok");

File: tests/independent_clients/python/call_convert.py
#!/usr/bin/env python3
"""Independent Python MCP client against a Control-backed Capability Gateway.

Uses only the public Streamable HTTP endpoint and bearer token; no shared
package filesystem with the Use host.
"""

from __future__ import annotations

import asyncio
import os
import sys

import httpx
from mcp import ClientSession
from mcp.client.streamable_http import streamable_http_client


async def main() -> int:
    endpoint = os.environ.get("A3S_GATEWAY_ENDPOINT")
    token = os.environ.get("A3S_GATEWAY_TOKEN")
    if not endpoint or not token:
        print(
            "A3S_GATEWAY_ENDPOINT and A3S_GATEWAY_TOKEN are required",
            file=sys.stderr,
        )
        return 2

    headers = {"Authorization": f"Bearer {token}"}
    async with httpx.AsyncClient(headers=headers) as http_client:
        async with streamable_http_client(endpoint, http_client=http_client) as (
            read_stream,
            write_stream,
            _,
        ):
            async with ClientSession(read_stream, write_stream) as session:
                await session.initialize()
                listed = await session.list_tools()
                names = [tool.name for tool in listed.tools]
                if names != ["convert"]:
                    print(f"unexpected tools: {names!r}", file=sys.stderr)
                    return 1
                result = await session.call_tool(
                    "convert", arguments={"args": []}
                )
                if result.isError:
                    print(f"call_tool failed: {result!r}", file=sys.stderr)
                    return 1
                structured = result.structuredContent or {}
                if structured.get("exitCode") != 0:
                    print(
                        f"unexpected structured content: {structured!r}",
                        file=sys.stderr,
                    )
                    return 1
    print("ok")
    return 0


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
